lib/change_stream.js

```diff
@@ -476,96 +475,73 @@ function waitForTopologyConnected(topology, options, callback) {
   }, 3000); // this is an arbitrary wait time to allow SDAM to transition
 }
 
-// Handle new change events. This method brings together the routes from the callback, event emitter, and promise ways of using ChangeStream.
-function processNewChange(args) {
-  const changeStream = args.changeStream;
-  const error = args.error;
-  const change = args.change;
-  const callback = args.callback;
-  const eventEmitter = args.eventEmitter || false;
-  const cursor = changeStream.cursor;
-
-  // If the cursor is null, then it should not process a change.
-  if (cursor == null) {
-    // We do not error in the eventEmitter case.
-    if (eventEmitter) {
-      return;
-    }
+// Wait for topology to re-connect, then create a new cursor.
+// If in eventEmitter mode, returns null and doesn't iterate the change stream.
+function resumeFromError({ changeStream, callback, eventEmitter }) {
+  const { cursor } = changeStream;
+  const {
+    options: { readPreference },
+    resumeOptions
+  } = cursor;
 
-    const error = new MongoError('ChangeStream is closed');
-    return typeof callback === 'function'
-      ? callback(error, null)
-      : changeStream.promiseLibrary.reject(error);
-  }
+  // stop listening to all events from old cursor
+  ['data', 'close', 'end', 'error'].forEach(event => cursor.removeAllListeners(event));
 
-  const topology = changeStream.topology;
-  const options = changeStream.cursor.options;
-  const wireVersion = maxWireVersion(cursor.server);
+  // close internal cursor, ignore errors
+  cursor.close();
 
-  if (error) {
-    if (isResumableError(error, wireVersion) && !changeStream.attemptingResume) {
-      changeStream.attemptingResume = true;
-
-      // stop listening to all events from old cursor
-      ['data', 'close', 'end', 'error'].forEach(event =>
-        changeStream.cursor.removeAllListeners(event)
-      );
+  // throw it away
+  changeStream.cursor = null;
 
-      // close internal cursor, ignore errors
-      changeStream.cursor.close();
+  // attempt resume
+  const resumeAttempt = maybePromise(callback, cb => {
```

Comment:
`maybePromise` is meant to live at the top level of an API, everything beneath it should explicitly be using callbacks IMO.